### DISCUSSION TITLE
Analyzer problems

### DISCUSSION THREAD

**Discussion by twollgam** [performous/performous#794] [Show and tell]
Writing some tests for the Analyzer I run in troubles. For the first tests I let analyse a calculated sine wave for guitar string frequencies. 4 of 8 tests fail. In three cases the half frequency is detected, one case detects 92Hz where 146Hz was expected. 
```cpp
    auto&& analyzer = Analyzer(48000., "guitar");
    
    fill(analyzer, 110.f);
    
    auto chordprovider = ChordProvider(analyzer);
    auto const result = chordprovider.getChord().getFrequencies();
    
    ASSERT_THAT(result, Not(IsEmpty()));
    EXPECT_THAT(*result.begin(), FloatNear(110, 2.0));
```
Output:
```
Value of: *result.begin()
Expected: is approximately 110 (absolute error <= 2)
  Actual: 55.6489 (of type float), which is -54.3511 from 110
```
Two frequencies:
```cpp
    auto&& analyzer = Analyzer(48000., "guitar", 100);
    
    fill(analyzer, 110.f, 82.41f);
    
    auto chordprovider = ChordProvider(analyzer);
    auto const result = chordprovider.getChord().getFrequencies();
    
    EXPECT_THAT(result, Contains(FloatNear(110.0f, 2.0)));
    EXPECT_THAT(result, Contains(FloatNear(82.41f, 2.0)));
```
Output:
```
Expected: contains at least one element that is approximately 110 (absolute error <= 2)
  Actual: { 95.6251, 184.205 }
/home/twollgam/devel/performous/performous/German/performous/testing/chordprovidertest.cc:133: Failure
Value of: result
Expected: contains at least one element that is approximately 82.410004 (absolute error <= 2)
  Actual: { 95.6251, 184.205 }
```

@Tronic can you explain this behavior? What can I do?

**Comment by twollgam** [comment]
If I increase the fft window size the results are good:

With `static const unsigned FFT_P = 12;` all tests pass. @Tronic is this a solution or are there any drawbacks?

**Reply by Baklap4** [reply]
Might wanna check the pitchdetector PR aswell: https://github.com/performous/performous/pull/541

**Comment by Tronic** [comment]
@twollgam The drawbacks of longer FFT window are longer latency (only relevant in realtime use as in Performous) and less reliability when the tones are changing over time (which apparently is not the case in your tests but which would be an issue for singing, pitch bending etc). 11 might be ideal for non-realtime applications, I would not go much longer than that due to the latter issue. The FFT step should be adjusted accordingly, try values around 1/4 of FFT length, maybe a bit less (e.g. Performous uses 200 sample step with FFT of 1024 samples).

And I recommend checking with the pitchdetector2 branch which uses different algorithm for combining the harmonics. Neither the one in Performous master nor the one used there is perfect (meaning it would be easy to implement a more intelligent matching of peaks into tones, perhaps at the expense of heavier CPU load), but if you are getting half frequencies then especially that might help because the main difference with pitchdetector2 is the harmonics matching.

Notice that the pitchdetector2 branch can also provide log messages with `performous --log pitch`, and for more detailed output uncomment in code the line that prints the peaks (i.e. all detected frequencies) and not only the tones (after harmonics are combined), to better understand why a wrong fundamental frequency is found. Paste the relevant part of those messages here if you need my analysis (notice that the peaks like comes right before the corresponding tones line in log).

**Comment by Tronic** [comment]
Re your tests, it is not entirely clear what `fill` and `chordprovider` are doing. Notice that the analyzer needs enough audio to produce the `m_tones` output (about 2000 samples with the default settings I guess). Also, if your chord frequencies are too close together, they might get accidentally merged, in particular when they fall within the same/nearby FFT bins (thus why the increase of FFT size would help) - this would appear on peaks output with "averaged" or errant frequencies rather than the pure ones. The peaks and tones merging code have a rather strict limit (half a semitone, IIRC), to reduce the chances of such errant merging occurring.

**Reply by Tronic** [reply]
A real guitar (just as other instruments, singing etc) would contain not only the fundamental frequency as in those tests, but also many harmonics. This is also the case with the Performous built-in synth mode (Ctrl+S while singing), which seems to be detected quite perfectly by the pitchdetector2 branch.
